File: sunnad-ios/sunnad-ios/Domain/Models/Habit.swift
import Foundation

enum HabitType: String, Codable, Hashable, Sendable {
    case binary
    case dhikr
}

enum HabitCategoryValue: String, Codable, Hashable, Sendable {
    case spiritual
    case physical
    case social
    case financial
    case learning
    case family
}

struct HabitReminder: Codable, Hashable, Sendable {
    var hour: Int
    var minute: Int

    init(hour: Int, minute: Int) {
        self.hour = min(max(hour, 0), 23)
        self.minute = min(max(minute, 0), 59)
    }

    init(date: Date, calendar: Calendar = .current, timeZone: TimeZone = .current) {
        var calendar = calendar
        calendar.timeZone = timeZone
        let components = calendar.dateComponents([.hour, .minute], from: date)
        self.hour = components.hour ?? 9
        self.minute = components.minute ?? 0
    }

    func asDateComponents() -> DateComponents {
        DateComponents(hour: hour, minute: minute)
    }
}

struct Habit: Identifiable, Codable, Hashable, Sendable {
    let id: UUID
    var name: String
    var icon: String
    var category: HabitCategoryValue
    var type: HabitType
    var targetCount: Int?
    var schedule: HabitSchedule
    var reminder: HabitReminder?
    var sortOrder: Int
    var archived: Bool
    var createdAt: Date
    var updatedAt: Date

    init(
        id: UUID = UUID(),
        name: String,
        icon: String,
        category: HabitCategoryValue = .spiritual,
        type: HabitType,
        targetCount: Int? = nil,
        schedule: HabitSchedule = .daily,
        reminder: HabitReminder? = nil,
        sortOrder: Int = 0,
        archived: Bool = false,
        createdAt: Date = Date(),
        updatedAt: Date = Date()
    ) {
        self.id = id
        self.name = name
        self.icon = icon
        self.category = category
        self.type = type
        self.targetCount = targetCount
        self.schedule = schedule
        self.reminder = reminder
        self.sortOrder = sortOrder
        self.archived = archived
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }

    var isDhikr: Bool {
        type == .dhikr
    }

    func isDue(on date: Date, calendar: Calendar = .current, timeZone: TimeZone = .current) -> Bool {
        !archived && schedule.isDue(on: date, calendar: calendar, timeZone: timeZone)
    }

    var normalizedTargetCount: Int {
        switch type {
        case .binary:
            return 1
        case .dhikr:
            return max(targetCount ?? 1, 1)
        }
    }
}

File: sunnad-android/app/src/main/java/com/arystan/almasuly/sunnadandroid/core/model/Habit.kt
package com.arystan.almasuly.sunnadandroid.core.model

import java.time.LocalDateTime
import java.util.UUID

enum class HabitType {
    BINARY,
    DHIKR
}

enum class HabitCategoryValue {
    SPIRITUAL,
    PHYSICAL,
    SOCIAL,
    FINANCIAL,
    LEARNING,
    FAMILY
}

data class HabitReminder(
    val hour: Int,
    val minute: Int
) {
    val normalizedHour: Int = hour.coerceIn(0, 23)
    val normalizedMinute: Int = minute.coerceIn(0, 59)
}

data class Habit(
    val id: UUID = UUID.randomUUID(),
    val name: String,
    val icon: String,
    val category: HabitCategoryValue = HabitCategoryValue.SPIRITUAL,
    val type: HabitType,
    val targetCount: Int? = null,
    val schedule: HabitSchedule = HabitSchedule.Daily,
    val reminder: HabitReminder? = null,
    val selectedDhikrKey: String = DEFAULT_DHIKR_KEY,
    val dhikrCountsByKey: Map<String, Int> = emptyMap(),
    val sortOrder: Int = 0,
    val archived: Boolean = false,
    val createdAt: LocalDateTime = LocalDateTime.now(),
    val updatedAt: LocalDateTime = LocalDateTime.now()
) {
    val normalizedTargetCount: Int
        get() = when (type) {
            HabitType.BINARY -> 1
            HabitType.DHIKR -> (targetCount ?: 1).coerceAtLeast(1)
        }

    val isDhikr: Boolean
        get() = type == HabitType.DHIKR

    fun isDue(on: java.time.LocalDate): Boolean = !archived && schedule.isDue(on)

    companion object {
        const val DEFAULT_DHIKR_KEY = "dhikr.choice.subhanallah"
    }
}
